File: Loss.cpp
#include "Loss.h"

Loss::Loss() {
	network = NULL;
};
Loss::Loss(BaseNetwork* network) : network(network) {};
void Loss::backpropagate() {
	if (network != NULL) {
		network->seq.backpropagate(error);
	}
};
Matrix<double> Loss::operator()(Matrix<double> output, Matrix<double> reference) {
	if (output.shape == reference.shape) {
		error = reference - output;
		return error;
	}
};

File: loss.py
from copy import copy
from matrix import Matrix

class Loss:
    def __init__(self, network):
        self.network = network

    def __call__(self, output: Matrix, reference: Matrix) -> Matrix:
        if (reference.shape == output.shape):
            output = copy(output)
            reference = copy(reference)
            self.error = reference - output
    
    def backpropagate(self):
        self.network.backpropagate(self.error)
        
    
        
